rosetta-lang/src/main/java/com/regnosys/rosetta/generator/java/types/JavaTypeTranslator.java

```diff
@@ -236,36 +215,50 @@ public JavaClass<?> toJavaType(RAttribute attr) {
 		}
 		return itemType;
 	}
+	public JavaClass<?> toJavaType(RFeature feature) {
+		if (feature instanceof RAttribute) {
+			return toJavaType((RAttribute) feature);
+		} else if (feature instanceof RMetaAttribute) {
+			return toItemJavaType((RMetaAttribute) feature);
+		} else {
+			throw new UnsupportedOperationException("No JavaType exists for feature: " + feature.getName());
+		}
+	}
 	public JavaClass<?> operationToReferenceWithMetaType(Operation op) {
-		Attribute attr;
+		RosettaFeature feature;
 		if (op.getPath() == null) {
-			attr = (Attribute)op.getAssignRoot(); // TODO: this won't work when assigning to an alias
+			feature = (Attribute)op.getAssignRoot(); // TODO: this won't work when assigning to an alias
 		} else {
 			List<Segment> segments = op.pathAsSegmentList();
-			attr = segments.get(segments.size() - 1).getAttribute();
+			feature = segments.get(segments.size() - 1).getFeature();
 		}
-		return toJavaReferenceType(typeProvider.getRTypeOfSymbol(attr));
+		return toJavaReferenceType(typeProvider.getRTypeOfFeature(feature, null));
 	}
-	
 	public JavaReferenceType operationToJavaType(ROperation op) {
-		RAttribute attr;
+		RFeature feature;
 		if (op.getPathTail().isEmpty()) {
-			attr = (RAttribute)op.getPathHead(); // TODO: this won't work when assigning to an alias
+			feature = (RFeature)op.getPathHead(); // TODO: this won't work when assigning to an alias
 		} else {
-			List<RAttribute> segments = op.getPathTail();
-			attr = segments.get(segments.size() - 1);
+			List<RFeature> segments = op.getPathTail();
+			feature = segments.get(segments.size() - 1);
 		}
-		return toJavaType(attr);
+		return toJavaType(feature);
 	}
 	public JavaClass<?> operationToReferenceWithMetaType(ROperation op) {
-		RAttribute attr;
+		RFeature feature;
 		if (op.getPathTail().isEmpty()) {
-			attr = (RAttribute)op.getPathHead(); // TODO: this won't work when assigning to an alias
+			feature = (RAttribute)op.getPathHead(); // TODO: this won't work when assigning to an alias
 		} else {
-			List<RAttribute> segments = op.getPathTail();
-			attr = segments.get(segments.size() - 1);
+			List<RFeature> segments = op.getPathTail();
+			feature = segments.get(segments.size() - 1);
+		}
+		if (feature instanceof RAttribute) {
+			return toJavaReferenceType(((RAttribute)feature).getRMetaAnnotatedType());
+		} else if (feature instanceof RMetaAttribute) {
+			return toJavaReferenceType(((RMetaAttribute)feature).getRType());
+		} else {
+			throw new UnsupportedOperationException("No JavaReferenceType exists for feature: " + feature.getName());
```

Comment:
This is the kind of switches I would like to get rid of in the future. Good exercise to think of a better solution.